ee/experimental/incident_utils.py

```diff
@@ -1,10 +1,140 @@
+import os
+import logging
+
 import numpy as np
 import pandas as pd
 import networkx as nx
 
 from typing import List
+from openai import OpenAI
+
+from datetime import datetime, timedelta
+
+from fastapi import Depends
+
+from ee.experimental.note_utils import NodeCandidateQueue, NodeCandidate
+from ee.experimental.graph_utils import create_graph
+from ee.experimental.statistical_utils import get_alert_pmi_matrix
+
+from pusher import Pusher
+
+from keep.api.models.db.alert import Alert, Incident
+from keep.api.core.db import (
+    assign_alert_to_incident,
+    is_alert_assigned_to_incident,
+    get_last_alerts,
+    create_incident_from_dict,
+    update_incident_summary,
+)
+
+from keep.api.core.dependencies import (
+    AuthenticatedEntity,
+    AuthVerifier,
+    get_pusher_client,
+)
+
+from keep.api.core.db import (
+    assign_alert_to_incident,
+    create_incident_from_dict,
+    get_incident_by_id,
+    get_last_alerts,
+    get_last_incidents,
+    write_pmi_matrix_to_db,
+)
+
+logger = logging.getLogger(__name__)
+
+ALGORITHM_VERBOSE_NAME = "Basic correlation alrithm v0.2"
+
+
+def calculate_pmi_matrix(
+    ctx: dict | None,  # arq context
+    tenant_id: str,
+    upper_timestamp: datetime = datetime.now() - timedelta(seconds=60 * 60),
+    use_n_historical_alerts: int = 10e10,
+    sliding_window: int = 4 * 60 * 60,
+    stride: int = 60 * 60,
+) -> dict:
+    logger.info(
+        "Calculating PMI coefficients for alerts",
+        extra={
+            "tenant_id": tenant_id,
+        },
+    )
+    alerts=get_last_alerts(tenant_id, limit=use_n_historical_alerts, upper_timestamp=upper_timestamp) 
+    pmi_matrix = get_alert_pmi_matrix(alerts, 'fingerprint', sliding_window, stride)
+    write_pmi_matrix_to_db(tenant_id, pmi_matrix)
+    
+    return {"status": "success"}
+
+
+async def mine_incidents_and_create_objects(
+        ctx: dict | None,  # arq context
+        tenant_id: str,
+        alert_lower_timestamp: datetime = datetime.now() - timedelta(seconds=60 * 60 * 60),
```

Comment:
It's dangerous to calculate values right in the arguments, since value will be calculated once at code initialization stage

```python
def a(b=datetime.now()):
    print(b)
    
a()
2024-08-09 15:19:16.277626
a()
2024-08-09 15:19:16.277626
a()
2024-08-09 15:19:16.277626
```

Reply:
@GlebBerjoskin

Reply:
Thx! Addressed [here](https://github.com/keephq/keep/commit/76508eb1a91805d71e26669945b05157dbb96486)